src/homomorphic_signature/mod.rs

```diff
@@ -0,0 +1,90 @@
+use core::ops::Add;
+
+use crate::Error;
+use ark_serialize::CanonicalSerialize;
+use ark_std::hash::Hash;
+use ark_std::rand::Rng;
+
+pub mod ncs1;
+pub trait HomomorphicSignatureScheme {
+    type Parameters: Clone + Send + Sync;
+    type PublicKey: CanonicalSerialize + Hash + Eq + Clone + Default + Send + Sync;
+    type SecretKey: CanonicalSerialize + Clone + Default;
+    type Signature: Clone + Default + Send + Sync + Add<Self::Signature, Output = Self::Signature>;
+    type Message: Clone + Default + Send + Sync + Add<Self::Message, Output = Self::Message>;
+    type Weight;
+
+    // Create single g2 element and n g1
+    fn setup<R: Rng>(rng: &mut R, n: usize) -> Result<Self::Parameters, Error>;
+
+    fn keygen<R: Rng>(
+        pp: &Self::Parameters,
+        rng: &mut R,
+    ) -> Result<(Self::PublicKey, Self::SecretKey), Error>;
+
+    fn sign(
+        pp: &Self::Parameters,
+        sk: &Self::SecretKey,
+        tag: &[u8],
+        index: &[u8],
+        message: &[Self::Message],
+        //rng: &mut R, TODO: Research if this makes sense to have per message randomness in HSS signing
+    ) -> Result<Self::Signature, Error>;
+
+    fn verify(
+        pp: &Self::Parameters,
+        pk: &Self::PublicKey,
+        tag: &[u8],
+        index: &[u8],
+        message: &[Self::Message],
+        signature: &Self::Signature,
+    ) -> Result<bool, Error>;
+
+    fn verify_aggregate(
+        pp: &Self::Parameters,
+        pk: &Self::PublicKey,
+        tag: &[u8],
+        message: &[Self::Message],
+        weights: &[Self::Weight],
+        signature: &Self::Signature,
+    ) -> Result<bool, Error>;
+
+    fn evaluate(
+        signature: &[Self::Signature],
+        weights: &[Self::Weight]
+    ) -> Result<Self::Signature, Error>;
+}
+
+#[cfg(test)]
+mod test {
+    use crate::homomorphic_signature::{ncs1, *};
+    // use ark_ed_on_bls12_381::EdwardsProjective as Curve;
+    use ark_bn254::Bn254 as Curve;
+    use ark_std::{test_rng, vec::Vec};
+    use blake2::Blake2s;
+    use ark_ec::bn::Bn;
+    use ark_ec::pairing::Pairing;
+    use crate::ark_std::UniformRand;
+
+
+    fn single_sign_and_verify<S: HomomorphicSignatureScheme>(tag: &[u8], index: &[u8], message: &[S::Message]) {
+        let rng = &mut test_rng();
+        let n = message.len();
+        let parameters = S::setup(rng, n).unwrap();
+        let (pk, sk) = S::keygen(&parameters, rng).unwrap();
+        let sig = S::sign(&parameters, &sk, tag, index, message).unwrap();
+
+        assert!(S::verify(&parameters, &pk, tag, index, message, &sig).unwrap());
+    }
```

Comment:
1. not sure if I understand the split into the test and helper function in this case - especially that assert is done inside the helper
2. generally I prefer to separate tests from the code - but in here it is ok

Reply:
This test is written to match the style of the repo. There will also be other tests added. It seems to be convention in rust to put unit tests in the parent module file. Then Integration test will be in separate files. https://stackoverflow.com/questions/25107526/how-should-rust-unit-tests-be-organised

Reply:
@tkornuta-semiotic let me know if we can resolve this as is, or if we need additional changes.

Reply:
this is a minor - I have accepted the PR already ;) Resolving.